src/dali/plugin/input/rest/kraken.py

```diff
@@ -0,0 +1,414 @@
+_MARGIN: str = 'margin'
+_TRADE: str = 'trade'
+_ROLLOVER: str = 'rollover'
+_TRANSFER: str = 'transfer'
+_SETTLED: str = 'settled'
+_CREDIT: str = 'credit'
+_STAKING: str = 'staking'
+_SALE: str = 'sale'
+
+# Record Limits
+_TRADE_RECORD_LIMIT: int = 50
+
+
+class InputPlugin(AbstractCcxtInputPlugin):
+
+    __EXCHANGE_NAME: str = "kraken"
+    __PLUGIN_NAME: str = "kraken_REST"
+    __DEFAULT_THREAD_COUNT: int = 1
+    __CACHE_FILE: str = 'kraken.pickle'
+
+    def __init__(
+        self,
+        account_holder: str,
+        api_key: str,
+        api_secret: str,
+        native_fiat: str,
+        username: Optional[str] = None,
+        thread_count: Optional[int] = __DEFAULT_THREAD_COUNT,
+        use_cache: Optional[bool] = True,
+    ) -> None:
+        self.__api_key = api_key
+        self.__api_secret = api_secret
+
+        # We will have a default start time of July 27th, 2011 since Kraken Exchange officially launched on July 28th.
+        super().__init__(account_holder, datetime(2011, 7, 27, 0, 0, 0, 0), native_fiat, thread_count)
+        self.__username = username
+        self.__logger: logging.Logger = create_logger(f"{self.__EXCHANGE_NAME}")
+        self.__timezone = pytz.timezone('UTC')
+        self._initialize_client()
+        self._client.load_markets()
+        self.baseId_to_base = {value[_BASE_ID]: value[_BASE] for key, value in self._client.markets_by_id.items()}
+        self.baseId_to_base.update({'BSV': 'BSV'})
+        self.use_cache = use_cache
+
+    def exchange_name(self) -> str:
+        return self.__EXCHANGE_NAME
+
+    def plugin_name(self) -> str:
+        return self.__PLUGIN_NAME
+
+    def _initialize_client(self) -> kraken:
+        return kraken(
+            {
+                "apiKey": self.__api_key,
+                "enableRateLimit": True,
+                "secret": self.__api_secret,
+            }
+        )
+
+    @property
+    def _client(self) -> kraken:
+        super_client: Exchange = super()._client
+        if not isinstance(super_client, kraken):
+            raise Exception("Exchange is not instance of class kraken.")
+        return super_client
+
+    def _get_process_deposits_pagination_detail_set(self) -> Optional[AbstractPaginationDetailSet]:
+        pass
+
+    def _get_process_withdrawals_pagination_detail_set(self) -> Optional[AbstractPaginationDetailSet]:
+        pass
+
+    def _get_process_trades_pagination_detail_set(self) -> Optional[AbstractPaginationDetailSet]:
+        pass
+
+    def _process_gains(
+        self,
+        in_transactions: List[InTransaction],
+        out_transactions: List[OutTransaction],
+    ) -> None:
+        pass
+
+    def _gather_api_data(self):
+        import pickle
+        import os
+
+        cache_exists = os.path.exists(self.__CACHE_FILE)
+        if self.use_cache and cache_exists:
+            with open(self.__CACHE_FILE, 'rb') as file:
+                something = pickle.load(file)
+                return something
```

Comment:
Caching is already built into `abstract_input_plugin.py` which is what `abstract_ccxt_input_plugin.py` inherits from. [See the revelant code here](https://github.com/eprbell/dali-rp2/blob/fdbbed328f6741379ba6357c785ec8947dddf1ca/src/dali/abstract_input_plugin.py#L42-L56).

Reply:
Hey guys, just want to let you guys know if you don't hear from me its not because I've gone MIA but just hit a really bad fever. A little heads up. Once I feel better I'll make sure to incorporate the requested changes.

Reply:
No worries, I hope you get better soon!

Reply:
Take it easy and get better soon!

Reply:
Just as an FYI, I'm using the "eyes" emoji to denote I saw the comment. I was originally using the "thumbs up" emoji but since everyone uses it, the "eyes" emoji is easier to track.

Reply:
Hey all, I've just pushed the latest with PR comments incorporated.

Reply:
@macanudo527 Can you take a look at my changes to `mypy.ini` and `/src/stubs/ccxt.pyi`? I'm not familiar with these file types and took a best guess at the changes requested. I hope I got it somewhat accurate.

Reply:
@ndopencode I think those look good. Be sure to run the tests outlined in the [developer section](https://github.com/eprbell/dali-rp2/blob/main/README.dev.md#development-workflow) and see if they give you errors. Mypy might complain about your code and you will have to clean it up before we can merge.